File: android-twa/app/src/main/java/app/polycal/widgets/ComposeAppWidget.java
package app.polycal.widgets;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;

/**
 * Shared binder: header opens the TWA composer empty; field/send open the IME sheet
 * (PC-479). Layouts use TextView + shape ImageView so AppWidgetHost can inflate
 * (vectors/EditText in RemoteViews cause "Can't load widget" on many launchers).
 */
public abstract class ComposeAppWidget extends AppWidgetProvider {
    abstract String mode();
    abstract int layoutRes();
    abstract int headerId();
    abstract int inputId();
    abstract int sendId();
    abstract int rootId();

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            RemoteViews views = new RemoteViews(context.getPackageName(), layoutRes());
            views.setOnClickPendingIntent(headerId(), launchIntent(context, appWidgetId, true, 1));
            views.setOnClickPendingIntent(rootId(), launchIntent(context, appWidgetId, true, 2));
            views.setOnClickPendingIntent(inputId(), launchIntent(context, appWidgetId, false, 3));
            views.setOnClickPendingIntent(sendId(), launchIntent(context, appWidgetId, false, 4));
            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }

    private PendingIntent launchIntent(Context context, int appWidgetId, boolean skipInput, int requestSalt) {
        Intent intent = new Intent(context, QuickComposeActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.putExtra(QuickComposeActivity.EXTRA_MODE, mode());
        intent.putExtra(QuickComposeActivity.EXTRA_SKIP_INPUT, skipInput);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        int requestCode = appWidgetId * 10 + requestSalt + (WidgetUrls.MODE_NLP.equals(mode()) ? 5000 : 0);
        return PendingIntent.getActivity(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
        );
    }
}

File: android-widgets/app/src/main/java/app/polycal/widgets/ComposeAppWidget.kt
package app.polycal.widgets

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews

/**
 * Shared binder for the two compose widgets. Header/chrome opens the web composer
 * immediately (empty is fine). The field and send open a one-line activity because
 * OEM IMEs on homescreen EditTexts often drop the typed text.
 */
abstract class ComposeAppWidget : AppWidgetProvider() {
    abstract val mode: String
    abstract val layoutRes: Int
    abstract val headerId: Int
    abstract val inputId: Int
    abstract val sendId: Int
    abstract val rootId: Int

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray,
    ) {
        for (appWidgetId in appWidgetIds) {
            val views = RemoteViews(context.packageName, layoutRes)
            views.setOnClickPendingIntent(
                headerId,
                launchIntent(context, appWidgetId, skipInput = true, requestSalt = 1),
            )
            views.setOnClickPendingIntent(
                rootId,
                launchIntent(context, appWidgetId, skipInput = true, requestSalt = 2),
            )
            views.setOnClickPendingIntent(
                inputId,
                launchIntent(context, appWidgetId, skipInput = false, requestSalt = 3),
            )
            views.setOnClickPendingIntent(
                sendId,
                launchIntent(context, appWidgetId, skipInput = false, requestSalt = 4),
            )
            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }

    private fun launchIntent(
        context: Context,
        appWidgetId: Int,
        skipInput: Boolean,
        requestSalt: Int,
    ): PendingIntent {
        val intent = Intent(context, QuickComposeActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            putExtra(QuickComposeActivity.EXTRA_MODE, mode)
            putExtra(QuickComposeActivity.EXTRA_SKIP_INPUT, skipInput)
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        }
        val requestCode = appWidgetId * 10 + requestSalt + if (mode == WidgetUrls.MODE_NLP) 5_000 else 0
        return PendingIntent.getActivity(
            context,
            requestCode,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }
}
